src/coreclr/src/vm/threadsuspend.cpp

```diff
@@ -6140,26 +5940,20 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
         if (hr == ERROR_TIMEOUT)
             STRESS_LOG0(LF_SYNC, LL_INFO1000, "SysSuspension colission");
 
+        // debugger holds TS lock when synchronizing. We cannot be suspending for GC.
+        _ASSERTE(!g_pDebugInterface->IsSynchronizing() ||
+                 !(reason == ThreadSuspend::SUSPEND_FOR_GC || reason == ThreadSuspend::SUSPEND_FOR_GC_PREP));
+
         // If the debugging services are attached, then its possible
         // that there is a thread which appears to be stopped at a gc
         // safe point, but which really is not. If that is the case,
         // back off and try again.
 
         // If this is not the GC thread and another thread has triggered
-        // a GC, then we may have bailed out of SuspendRuntime, so we
-        // must resume all of the threads and tell the GC that we are
-        // at a safepoint - since this is the exact same behaviour
-        // that the debugger needs, just use it's code.
+        // a GC, then we yield by resuming all the threads and trying again.
         if ((hr == ERROR_TIMEOUT)
 #ifdef DEBUGGING_SUPPORTED
-             || (CORDebuggerAttached() &&
-            // When the debugger is synchronizing, trying to perform a GC could deadlock. The GC has the
-            // threadstore lock and synchronization cannot complete until the debugger can get the
-            // threadstore lock. However the GC can not complete until it sends the BeforeGarbageCollection
-            // event, and the event can not be sent until the debugger is synchronized. In order to break
-            // this deadlock cycle the GC must give up the threadstore lock, allow the debugger to synchronize,
-            // then try again.
-                 (g_pDebugInterface->ThreadsAtUnsafePlaces() || g_pDebugInterface->IsSynchronizing()))
```

Comment:
Since GC does call `BeforeGarbageCollection` I am less sure that `IsSynchronizing`  while suspending is an impossible state. I think I will just bring the check back.
It may be worth examining closely if it is needed. However we would do this check rarely and only if debugger is attached, so even if it is an impossible state, the check won't be a big deal.